### DISCUSSION TITLE
Certificate failing to validate on Windows, Openssl, MSQUIC

### DISCUSSION THREAD

**Discussion by cjgatto** [microsoft/msquic#2991] [Q&A]
I have a client key and cert for my msquic client, and a server key and cert for my msquic server. Certificates are trusted by Digicert CA.

I've compiled msquic with openssl. The msquic server runs on a Linux system, the client can run on Linux or Windows.

On Linux, the client and server authenticate just fine.

On Windows, the connections fails. The Linux server logs msquic status code [QUIC_STATUS_BAD_CERTIFICATE](https://github.com/microsoft/msquic/blob/main/src/inc/msquic_posix.h#L133), the Windows client logs a code of `0x8041012a` but that doesn't seem to correlate to a status in [msquic_winuser.h](https://github.com/microsoft/msquic/blob/main/src/inc/msquic_winuser.h). If I set `QUIC_CREDENTIAL_FLAG_NO_CERTIFICATE_VALIDATION` on the Windows client, the connection works.

When running my certs through `openssl.exe x509 -in sample.cert -text -noout`, the results look normal. The fact that these certs work fine on Linux-to-Linux systems suggests there isn't an issue with the certs themselves.

The only thing I can think of is the openssl built and linked by msquic isn't picking up any necessary Window's certificate chains. Is there something I can try to diagnose this further? Using CMake on Windows, I built msquic with `-DQUIC_TLS=openssl`.

**Comment by nibanks** [comment]
> Windows client logs a code of 0x8041012a but that doesn't seem to correlate to a status in [msquic_winuser.h](https://github.com/microsoft/msquic/blob/main/src/inc/msquic_winuser.h)

`0x8041012a` comes from a combination of:
```c
#define QUIC_TLS_ALERT_HRESULT_PREFIX   _HRESULT_TYPEDEF_(0x80410100L)
```
```c
#define QUIC_STATUS_TLS_ALERT(Alert)        (QUIC_TLS_ALERT_HRESULT_PREFIX | (0xff & Alert))
```
```c
#define QUIC_STATUS_BAD_CERTIFICATE         QUIC_STATUS_TLS_ALERT(42)   // Bad Certificate
```
Because `0x2a` equals `42`, so `QUIC_STATUS_TLS_ALERT(42)` equals `0x8041012a`.

Now, as to why Windows doesn't like the certificate, the Windows logs more information to ETW about the specific errors. My guess is that for some reason, Windows doesn't trust the root of the cert by default. Try grabbing the logs to find out more.

**Reply by cjgatto** [reply]
I believe the issue is validation fails due to the missing intermediate CA and root certificate. This is interesting because the .crt itself has all the pieces (leaf, intermediate, root) concatenated. 

On Windows, I can test the validity by telling openssl to inspect the cert for the rest of the chain
```
$ openssl.exe verify -CAfile client.crt client.crt
client.crt: OK
```

I've installed the intermediate and root certs to the system CA store, but my msquic application doesn't appear to be loading that.

Is there a way around this? I'd bundle and pass the certs in manually, but I see https://github.com/microsoft/msquic/issues/2820 is still open.

**Reply by nibanks** [reply]
Right now, you could just do custom certificate validation, and call the APIs you want, right?

**Reply by cjgatto** [reply]
So a combination of `QUIC_CONNECTION_EVENT_PEER_CERTIFICATE_RECEIVED` and `QUIC_CREDENTIAL_FLAG_NO_CERTIFICATE_VALIDATION`?

**Reply by nibanks** [reply]
Yes, I believe that should do the job.

**Reply by shashenli** [reply]
> hello i need some help would u pls help me? thank u. i use self signed certificate on server and i have already added the CA root certificate in Windows certMgr's trusted root signing certificate but still  0x8041012a i dont know why

**Reply by cjgatto** [reply]
Hi @shashenli, the suggestion above worked for me. The underlying issue is that OpenSSL's default behavior does not load certs installed on Windows, nor does MSQUIC provide an API to pass a CA for validation. 

The solution for me was to use `QUIC_CONNECTION_EVENT_PEER_CERTIFICATE_RECEIVED` with  `QUIC_CREDENTIAL_FLAG_NO_CERTIFICATE_VALIDATION`, catch the certificate received event in a callback, and use OpenSSL directly to perform validation with the received cert and your CA. 

However, I believe `0x8041012a` is a generic error for "Bad Certificates." Your certs might be failing in other ways. Would be best to check they are valid as well.

**Reply by shashenli** [reply]
Thank u for ur help i will try it

**Reply by shashenli** [reply]
seems like this is a useful way to solve it

**Reply by shashenli** [reply]
but if OpenSSL does not load certs installed on Windows where will OpenSSL load certs?

**Reply by cjgatto** [reply]
You'll have to find that in the OpenSSL docs, `X509_LOOKUP_load_file` for example loads a cert file to an in-memory cert store

**Reply by shashenli** [reply]
So if i run client on linux it will ask OS to load CA cert?but if i run client on windows it will not

**Reply by shashenli** [reply]
if i catch the certificate received event in a callback what should i do to perform validation with the received cert and my CA. would u pls show me how u do on code? ur help will be appreciated
